bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/DbLedgerStorage.java

```diff
@@ -0,0 +1,767 @@
+            if (entry != null) {
+                recordSuccessfulEvent(readCacheHitStats, startTime);
+                recordSuccessfulEvent(readEntryStats, startTime);
+                return entry;
+            }
+        } finally {
+            writeCacheMutex.readLock().unlock();
+        }
+
+        // Try reading from read-ahead cache
+        ByteBuf entry = readCache.get(ledgerId, entryId);
+        if (entry != null) {
+            recordSuccessfulEvent(readCacheHitStats, startTime);
+            recordSuccessfulEvent(readEntryStats, startTime);
+            return entry;
+        }
+
+        // Read from main storage
+        try {
+            long entryLocation = entryLocationIndex.getLocation(ledgerId, entryId);
+            if (entryLocation == 0) {
+                throw new NoEntryException(ledgerId, entryId);
+            }
+            entry = entryLogger.readEntry(ledgerId, entryId, entryLocation);
+            readCache.put(ledgerId, entryId, entry);
+
+            // Try to read more entries
+            long nextEntryLocation = entryLocation + 4 /* size header */ + entry.readableBytes();
+            fillReadAheadCache(ledgerId, entryId + 1, nextEntryLocation);
+
+            recordSuccessfulEvent(readCacheMissStats, startTime);
+            recordSuccessfulEvent(readEntryStats, startTime);
+            return entry;
+        } catch (NoEntryException e) {
+            recordFailedEvent(readEntryStats, startTime);
+            throw e;
+        }
+    }
+
+    private void fillReadAheadCache(long orginalLedgerId, long firstEntryId, long firstEntryLocation) {
+        try {
+            long firstEntryLogId = (firstEntryLocation >> 32);
+            long currentEntryLogId = firstEntryLogId;
+            long currentEntryLocation = firstEntryLocation;
+            int count = 0;
+            long size = 0;
+
+            while (count < readAheadCacheBatchSize && currentEntryLogId == firstEntryLogId) {
+                ByteBuf entry = entryLogger.internalReadEntry(orginalLedgerId, -1, currentEntryLocation);
+
+                long currentEntryLedgerId = entry.getLong(0);
+                long currentEntryId = entry.getLong(8);
+
+                if (currentEntryLedgerId != orginalLedgerId) {
+                    // Found an entry belonging to a different ledger, stopping read-ahead
+                    entry.release();
+                    return;
+                }
+
+                // Insert entry in read cache
+                readCache.put(orginalLedgerId, currentEntryId, entry);
+
+                count++;
+                size += entry.readableBytes();
+
+                currentEntryLocation += 4 + entry.readableBytes();
+                currentEntryLogId = currentEntryLocation >> 32;
+
+                entry.release();
+            }
+
+            readAheadBatchCountStats.registerSuccessfulValue(count);
+            readAheadBatchSizeStats.registerSuccessfulValue(size);
+        } catch (Exception e) {
```

Comment:
Is there a motivation for catching Exception and not a more specific set of exception?

Reply:
I think the intention was to ignore failures in read ahead, since we can already complete the current read entry operation.